src/MonoTorrent/MonoTorrent.Streaming/LocalStream.cs

```diff
@@ -0,0 +1,153 @@
+// 
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+// 
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+//
+
+
+using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+using MonoTorrent.Client;
+using MonoTorrent.Client.PiecePicking;
+
+namespace MonoTorrent.Streaming
+{
+    /// <summary>
+    /// A seekable Stream which can be used to access a <see cref="TorrentFile"/> while it is downloading.
+    /// If the stream seeks to a location which hasn't been downloaded yet, <see cref="Read(byte[], int, int)"/>
+    /// will block until the data is available. <see cref="ReadAsync(byte[], int, int, CancellationToken)"/>
+    /// will perform a non-blocking wait for the data.
+    /// </summary>
+    class LocalStream : Stream
+    {
+        long position;
+
+        public override bool CanRead => true;
+
+        public override bool CanSeek => true;
+
+        public override bool CanWrite => false;
+
+        public override long Length => File.Length;
+
+        public override long Position {
+            get => position;
+            set => Seek (value, SeekOrigin.Begin);
+        }
+
+        TorrentFile File { get; }
+
+        TorrentManager Manager { get; }
+
+        StreamingPiecePicker Picker { get; }
+
+        FileStream Stream { get; set; }
+
+        public LocalStream (TorrentManager manager, TorrentFile file, StreamingPiecePicker picker)
+        {
+            Manager = manager;
+            File = file;
+            Picker = picker;
+        }
+
+        protected override void Dispose (bool disposing)
+        {
+            base.Dispose (disposing);
+            if (disposing) {
+                Stream?.Dispose ();
+            }
+        }
+
+        public override void Flush ()
+            => throw new NotSupportedException ();
+
+        public override int Read (byte[] buffer, int offset, int count)
+            => ReadAsync (buffer, offset, count, CancellationToken.None).GetAwaiter ().GetResult ();
+
+        public override async Task<int> ReadAsync (byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            long torrentOffset = 0;
+            foreach (var file in Manager.Torrent.Files) {
+                if (file == File)
+                    break;
+
+                torrentOffset += file.Length;
+            }
+
+            var startPiece = (int) (torrentOffset + offset + Position) / Manager.Torrent.PieceLength;
```

Comment:
@alanmcgovern `torrentOffset + offset + Position` is duplicated. Can be moved to a variable like `startOffset`. https://github.com/ilyalatt/BitTorrentStorage/blob/master/src/BitTorrentStorage/BtsFileStream.cs#L92

Reply:
👍